src/Microsoft.TestPlatform.Common/Utilities/MetadataReaderHelper.cs

```diff
@@ -0,0 +1,156 @@
+
+                        // Get the custom attribute TypeDef handle
+                        var typeDefinitionHandle = methodDefinition.GetDeclaringType();
+
+                        // Read TypeDef metadata row
+                        var typeDef = metadataReader.GetTypeDefinition(typeDefinitionHandle);
+
+                        // Check the attribute type full name
+                        attributeFullName = $"{metadataReader.GetString(typeDef.Namespace)}.{metadataReader.GetString(typeDef.Name)}";
+                        if (attributeFullName == TestExtensionTypesAttributeV2)
+                        {
+                            // We don't do any signature verification to allow future possibility to add new parameter in a back compat way.
+                            // Get signature blob using methodDefinition.Signature index into Blob heap - ECMA-335 II.22.26 MethodDef : 0x06, 'Signature' column
+                            // BlobReader signatureReader = metadataReader.GetBlobReader(methodDefinition.Signature);
+                            // var decoder = new SignatureDecoder<string, object>(new SignatureDecoder(), metadataReader, genericContext: null);
+                            // var ctorDecodedSignature = decoder.DecodeMethodSignature(ref signatureReader);
+                            // Log the signature for analysis purpose
+                            // EqtTrace.Verbose($"MetadataReaderExtensionsHelper: Find possible good attribute candidate '{attributeFullName}' with ctor signature: '{ctorDecodedSignature.ReturnType}" +
+                            //    $" ctor ({(ctorDecodedSignature.ParameterTypes.Length > 0 ? ctorDecodedSignature.ParameterTypes.Aggregate((a, b) => $"{a},{b}").Trim(',') : "(")})'");
+
+                            // Read the ctor signature values - ECMA-335 II.23.3 Custom attributes
+                            BlobReader valueReader = metadataReader.GetBlobReader(customAttribute.Value);
+                            // Verify the prolog
+                            if (valueReader.ReadUInt16() == 1)
+                            {
+                                // Expected ctor shape ctor(string,string,System.Type,Int32)
+                                // [assembly: TestExtensionTypesV2(ExtensionMetadata.ExtensionType, ExtensionMetadata.Uri, typeof(ExtensionImplementation), 1)]
+
+                                // If the parameter kind is string, (middle line in above diagram) then the blob contains
+                                // a SerString – a PackedLen count of bytes, followed by the UTF8 characters.If the
+                                // string is null, its PackedLen has the value 0xFF(with no following characters).If
+                                // the string is empty(“”), then PackedLen has the value 0x00(with no following
+                                // characters).
+                                string extension = valueReader.ReadSerializedString();
```

Comment:
>extension [](http://example.com/codeflow?start=39&length=9)

what are the values that this can take?

Reply:
First two value for the attribute are for "future use".
We want to be able in future to understand from the attribute the "type" of extension and the "identifier" for better performance and to avoid to "get the type" if we're searching for a different type of extension(i.e. DataCollector, TestAdapter etc...).
We didn't define yet the precise "type name" for all extensions.

Reply:
For now we're using Type and version.

Reply:
Could we add that as a comment to code too please?